flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/fs/FsWatchService.java

```diff
@@ -0,0 +1,90 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+
+import static java.nio.file.StandardWatchEventKinds.ENTRY_CREATE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+
+/** Service which is able to watch local filesystem directories. */
+public class FsWatchService extends Thread {
+    private static final Logger LOG = LoggerFactory.getLogger(FsWatchService.class);
+
+    private final String directoryPath;
+
+    public FsWatchService(String directoryPath) {
+        this.directoryPath = directoryPath;
+    }
+
+    @Override
+    public void run() {
+        try (WatchService watcher = FileSystems.getDefault().newWatchService()) {
+            LOG.info("Starting watching path: " + directoryPath);
+            Path realDirectoryPath = Paths.get(directoryPath).toRealPath();
+            LOG.info("Path is resolved to real path: " + realDirectoryPath);
+            realDirectoryPath.register(watcher, ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY);
+            onWatchStarted(realDirectoryPath);
+
+            while (true) {
+                LOG.debug("Taking watch key");
+                WatchKey watchKey = watcher.take();
+                LOG.debug("Watch key arrived");
+                for (WatchEvent<?> watchEvent : watchKey.pollEvents()) {
+                    LOG.debug("Watch event count: " + watchEvent.count());
+                    if (watchEvent.kind() == OVERFLOW) {
+                        LOG.error("Filesystem events may have been lost or discarded");
+                        Thread.yield();
+                    } else if (watchEvent.kind() == ENTRY_CREATE) {
+                        onFileOrDirectoryCreated((Path) watchEvent.context());
+                    } else if (watchEvent.kind() == ENTRY_DELETE) {
+                        onFileOrDirectoryDeleted((Path) watchEvent.context());
+                    } else if (watchEvent.kind() == ENTRY_MODIFY) {
+                        onFileOrDirectoryModified((Path) watchEvent.context());
+                    } else {
+                        throw new IllegalStateException("Invalid event kind: " + watchEvent.kind());
+                    }
+                }
+                watchKey.reset();
+            }
+        } catch (InterruptedException e) {
+            LOG.info("Filesystem watcher interrupted");
+        } catch (Exception e) {
+            LOG.error("Filesystem watcher received exception and stopped: " + e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    protected void onWatchStarted(Path realDirectoryPath) {}
+
+    protected void onFileOrDirectoryCreated(Path relativePath) {}
```

Comment:
Could the `onFileOrDirectoryCreated `, `onFileOrDirectoryDeleted` and `onFileOrDirectoryModified` merge into one interface `onWatchedEvent(WatchEvent.Kind kind, Path relativePath)`?

Reply:
I think it's internal detail that the internal implementation uses `WatchService` which is sending `WatchEvent.Kind` type which shouldn't appear in the API of the class. Simply loose couple.